### PR TITLE
JAVA-5531 :sparkles: Create target directory in contrast-maven-plugin when running mvn clean

### PR DESCRIPTION
**Pull request by ammentorp**
As of version 2.13, running `mvn clean` in the `contrast-maven-plugin` will currently remove the target directory.  This is causing exceptions to be thrown when it fails to save the agent to the directory. This PR remedies this issue so now `mvn clean` successfully generates the target directory if not already present.

### CODE REVIEW

**Review by tylernhoward — APPROVED** [review]

**Comment by tylernhoward** [comment]
No possible unit test?

**Review by gilday — CHANGES_REQUESTED** [review]
Before we reach for `mkdir -p` to make this problem go away, I want us to consider more carefully why the `target` directory does not already exist by the time the plugin's `install` goal runs.

If you take a look at the Maven [Lifecycle Reference](https://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html#Lifecycle_Reference), it will show you that `initialize` phase is responsible for creating directories, and I assume this includes the `project.build.directory`. This plugin's `install` goal is set to use the `validate` lifecycle phase by default, so it makes sense why the directory has not yet been created.

But why are we tying this plugin to the `validate` phase? I'm not getting much from the git history for the phase configuration. I'd instead look to the [maven-dependency-plugin's copy goal](https://maven.apache.org/plugins/maven-dependency-plugin/copy-mojo.html) for guidance on what phase the `install` goal should bind to (i.e. `process-sources`): our `install` goal is similar to `maven-dependency-plugin:copy` in that it retrieves an artifact and copies it into a directory in the `project.build.directory`.

So, if you change the default phase to `process-sources`, does that solve this problem without violating Maven norms around directory creation?

**Comment by ammentorp** [comment]
@gilday @tylernhoward 
Changing the default phase to `process-sources` doesn't solve this problem. I tooled around with changing the default phase to `generate-sources`, `install`, and `initialize` to try and force the creation of this directory. No such luck.

I did some digging into the git history between the 2.12 and 2.13 release and found the `installJavaAgent` method used to be implemented as such:
```    
      // Save the jar to the 'target' directory
      agentFile = new File(project.getBuild().getDirectory() + File.separator + AGENT_NAME);

      try {
        FileUtils.writeByteArrayToFile(agentFile, javaAgent);
      } catch (IOException e) {
        throw new MojoExecutionException("Unable to save the latest java agent.", e);
      }
```

Specifically, stepping through the call to `FileUtils.writeByteArrayToFile(agentFile, javaAgent);` we find that this indirectly creates the parent directories of the `agentFile` if one doesn't already exist. It was a messy deep dive, but I ultimately found that `PathUtils.newOutputStream` is called down the line, which is defined here: https://github.com/apache/commons-io/blob/b1842c0e0c5af94976ce6504c78e7ef6200abec9/src/main/java/org/apache/commons/io/file/PathUtils.java#L1168

All this is to say, that by checking if the parent directory exists in `installJavaAgent` is simply reverting back to its initial functionality.

**Comment by gilday** [comment]
> All this is to say, that by checking if the parent directory exists in `installJavaAgent` is simply reverting back to its initial functionality.

I understand, but I still wanted to question the validity of the initial functionality. If we're following maven-dependency-plugin's example, then we should do both: change the default phase binding to something more appropriate _and_ use a utility to force the creation of parent directories.

The `maven-dependency-plugin` [uses Plexus Utils to copy files](https://github.com/apache/maven-dependency-plugin/blob/e52bc0248c00dbf5458a0ce080db260148dab4b9/src/main/java/org/apache/maven/plugins/dependency/AbstractDependencyMojo.java#L191) and these utilities automatically create missing directories. Reusing Plexus Utils is wise, if only to make sure our `IOException` handling is consistent with other Maven plugins:

```java
    // Save the jar to the 'target' directory
    final Path target = Paths.get(project.getBuild().getDirectory());
    try {
      FileUtils.forceMkdir(target.toFile());
    } catch (final IOException e) {
      throw new MojoFailureException("Unable to create directory " + target, e);
    }
    final Path agent = target.resolve(AGENT_NAME);
```

Finally, we can get one of the end-to-end tests to fail if we remove the phase binding override from the POM:

```diff
diff --git a/src/test/resources/it/spring-boot/pom.xml b/src/test/resources/it/spring-boot/pom.xml
index ca5cf52..9c4715b 100644
--- a/src/test/resources/it/spring-boot/pom.xml
+++ b/src/test/resources/it/spring-boot/pom.xml
@@ -64,7 +64,6 @@
             <goals>
               <goal>install</goal>
             </goals>
-            <phase>pre-integration-test</phase>
           </execution>
         </executions>
         <configuration>
```

File: src/test/java/com/contrastsecurity/maven/plugin/it/ContrastInstallAgentMojoIT.java [hunk 1]
```diff
@@ -55,6 +55,7 @@ public void test(final ContrastAPI contrast) throws IOException, VerificationExc
     final Verifier verifier = Verifiers.springBoot(contrast.connection());
 
     // WHEN execute the "verify" goal
+    verifier.executeGoal("clean");
```
Comment (gilday) [inline src/test/java/com/contrastsecurity/maven/plugin/it/ContrastInstallAgentMojoIT.java]: This isn't needed. You can still reproduce the error without it.

File: src/main/java/com/contrastsecurity/maven/plugin/ContrastInstallAgentMojo.java [hunk 2]
```diff
@@ -387,8 +388,21 @@ Path installJavaAgent(ContrastSDK connection) throws MojoFailureException {
...
+      FileUtils.forceMkdir(target.toFile());
+    } catch (final IOException e) {
+      throw new MojoFailureException("Unable to create directory " + target, e);
+    }
+    final Path agent = target.resolve(AGENT_NAME);
+    try {
+      Path parentDir = agent.getParent().getFileName();
+
+      // If 'target' directory doesn't exist, create it
+      if (!Files.exists(parentDir)) {
```
Comment (gilday) [inline src/main/java/com/contrastsecurity/maven/plugin/ContrastInstallAgentMojo.java]: This is redundant: you've already created the directories on line 393. You don't need any of the changes after line 397, right?

**Review by gilday — CHANGES_REQUESTED** [review]

File: src/main/java/com/contrastsecurity/maven/plugin/ContrastInstallAgentMojo.java (hunk 2 above)
Comment (ammentorp) [reply src/main/java/com/contrastsecurity/maven/plugin/ContrastInstallAgentMojo.java]: Agh, the patch got messed up :shaking-angry-fist:

**Review by gilday — APPROVED** [review]
lgtm!

**Review by tylernhoward — APPROVED** [review]

**Pull request merged**